File: Benchmarks/Core/Systems/AnimationSystem2D.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Systems.Contract;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems
{
    /// <summary>
    /// Sistema encargado de gestionar todas las animaciones de las entidades.
    /// Replica el patron de iteracion con ref mutation del engine.
    /// </summary>
    public class AnimationSystem2D : ISystem
    {
        private EventManager EventManager { get; set; }

        public override void InitializeSystem(EntityManager @entities)
        {
            EventManager = EventManager.Instance;
            EntityManager = @entities;
        }

        public override void UpdateSystem(GameTime gameTime)
        {
            Span<AnimationComponent2D> animationsComponents = GetAnimationComponentsAsSpan();
            Span<SpriteComponent2D> spriteComponents = GetSpriteComponentsAsSpan();
            if (EntityManager == null) return;
            foreach (ref Yotsuba entity in GetEntitiesAsSpan())
            {
                if (!entity.HasComponent(YTBComponent.Animation) || !entity.HasComponent(YTBComponent.Sprite)) continue;
                ref AnimationComponent2D animationComponent = ref animationsComponents[entity.Id];
                ref SpriteComponent2D spriteComponent = ref spriteComponents[entity.Id];

                if (animationComponent.CurrentAnimationType.Item2.IsFinished)
                {
                    if (!animationComponent.CurrentAnimationType.Item2.FinishedWasMarked)
                    {
                        animationComponent.CurrentAnimationType.Item2.FinishedWasMarked = true;
                    }
                }
                else
                {
                    Rectangle region = animationComponent.CurrentAnimationType.Item2.CurrentFrameRect(gameTime);
                    spriteComponent.TextureId = animationComponent.CurrentAnimationType.Item2.TextureId;
                    spriteComponent.SourceRectangle = region;
                }
            }
        }

        public override void SharedEntityForEachUpdate(ref Yotsuba Entidad, GameTime time) { }
        public override void SharedEntityInitialize(ref Yotsuba Entidad) { }
        public override void Dispose() { }
    }
}

File: Benchmarks_FSharp/Core/Systems/AnimationSystem2D.fs
namespace Benchmarks.Core.Systems

open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type AnimationSystem2D() =
    inherit ISystem()

    let mutable eventManager: EventManager = Unchecked.defaultof<EventManager>

    override this.InitializeSystem(entities: EntityManager) =
        eventManager <- EventManager.Instance
        this.EntityManager <- entities

    override this.UpdateSystem(gameTime: GameTime) =
        let animationsComponents = this.GetAnimationComponentsAsSpan()
        let spriteComponents = this.GetSpriteComponentsAsSpan()
        if isNull this.EntityManager then ()
        else
        let entities = this.GetEntitiesAsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            if entity.HasComponent(YTBComponent.Animation) && entity.HasComponent(YTBComponent.Sprite) then
                let animationComponent = &animationsComponents.[entity.Id]
                let spriteComponent = &spriteComponents.[entity.Id]
                let struct (_, animData) = animationComponent.CurrentAnimationType
                if animData.IsFinished then
                    if not animData.FinishedWasMarked then
                        let mutable current = animationComponent.CurrentAnimationType
                        let struct (at, ad) = current
                        let mutable ad2 = ad
                        ad2.FinishedWasMarked <- true
                        animationComponent.CurrentAnimationType <- struct (at, ad2)
                else
                    let mutable current = animationComponent.CurrentAnimationType
                    let struct (at, ad) = current
                    let mutable ad2 = ad
                    let region = ad2.CurrentFrameRect(gameTime)
                    spriteComponent.TextureId <- ad2.TextureId
                    spriteComponent.SourceRectangle <- region
                    animationComponent.CurrentAnimationType <- struct (at, ad2)

    override _.SharedEntityForEachUpdate(_, _) = ()
    override _.SharedEntityInitialize(_) = ()
    override _.Dispose() = ()
